backend/Tim.Backend/Providers/Database/CouchbaseDbClient.cs

```diff
@@ -0,0 +1,179 @@
+    using System.Threading.Tasks;
+    using Couchbase;
+    using Couchbase.KeyValue;
+    using Couchbase.Management.Buckets;
+    using Couchbase.Management.Collections;
+    using Couchbase.Management.Query;
+    using Serilog;
+    using Tim.Backend.Models;
+    using Tim.Backend.Models.KustoQuery;
+    using Tim.Backend.Models.Templates;
+    using Tim.Backend.Startup.Config;
+
+    /// <summary>
+    /// Manages the connection and initialization for the couchbase database.
+    /// </summary>
+    public class CouchbaseDbClient : IDatabaseClient
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CouchbaseDbClient"/> class.
+        /// </summary>
+        /// <param name="dbConfigs">Collection/table names.</param>
+        public CouchbaseDbClient(DatabaseConfiguration dbConfigs)
+        {
+            Logger = Log.Logger;
+
+            Configs = dbConfigs;
+        }
+
+        /// <summary>
+        /// Gets or sets the Couchbase db client.
+        /// </summary>
+        public ICluster CouchBaseClient { get; set; }
+
+        /// <summary>
+        /// Gets or sets the database configs.
+        /// </summary>
+        public DatabaseConfiguration Configs { get; set; }
+
+        /// <summary>
+        /// Gets or sets the bucket after initialization.
+        /// </summary>
+        public IBucket Bucket { get; set; }
+
+        /// <summary>
+        /// Gets the logger.
+        /// </summary>
+        protected ILogger Logger { get; }
+
+        /// <summary>
+        /// Generate the collection name for this entity type.
+        /// </summary>
+        /// <typeparam name="T">Entity type.</typeparam>
+        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
+        public static string GetCollectionName<T>()
+        {
+            return typeof(T).Name;
+        }
+
+        /// <inheritdoc/>
+        public async Task Connect()
+        {
+            Logger.Information($"CouchbaseDbClient is connecting to database.");
+            CouchBaseClient = await Cluster.ConnectAsync(
+                Configs.DatabaseConnection,
+                new ClusterOptions
+                {
+                    UserName = Configs.DbUserName,
+                    Password = Configs.DbUserPassword,
+                    KvTimeout = TimeSpan.FromSeconds(10),
+                });
+            Logger.Information($"CouchbaseDbClient has been created.");
+        }
+
+        /// <inheritdoc/>
+        public async Task Initialize()
+        {
+            if (CouchBaseClient is null)
+            {
+                throw new InvalidOperationException("CouchBaseClient must be initialized with a connection.");
+            }
+
+            await CouchBaseClient.WaitUntilReadyAsync(TimeSpan.FromMinutes(2));
+            Bucket = await CreateBucketIfNotExists(Configs.BucketName);
+            await CreateCollectionIfNotExists<KustoQueryRun>();
```

Comment:
I like how this is done but worry that something like this can be easy to forget when adding a new db class

Reply:
Yeah agreed. I couldn't come up with a fancy way to automatically register collections for initialization. I think it's doable as part of when the object is registered in the services, but this is already too late.